### PR TITLE
colpali v1.3 by AndrewOgn

### PR DESCRIPTION
**Pull request by joein**
it's a draft of second iteration of work on colpali #394

### CODE REVIEW

File: fastembed/late_interaction_multimodal/colpali.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    OnnxMultimodalModel,
+    TextEmbeddingWorker,
+    ImageEmbeddingWorker,
+)
+
+
+supported_colbert_models = [
+    {
+        "model": "colpali",
+        "dim": ...,
```
Comment (I8dNLo) [inline fastembed/late_interaction_multimodal/colpali.py:22]: It should contain dim of outputs

File: fastembed/late_interaction_multimodal/colpali.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+            self.load_onnx_model()
+
+    @classmethod
+    def list_supported_models(cls) -> list[dict[str, Any]]:
+        """Lists the supported models.
+
+        Returns:
+            list[dict[str, Any]]: A list of dictionaries containing the model information.
+        """
+        return supported_colbert_models
```
Comment (I8dNLo) [inline fastembed/late_interaction_multimodal/colpali.py:113]: supported_colpali_models

File: fastembed/late_interaction_multimodal/onnx_multimodal_model.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+        """
+        return onnx_input
+
+    def _preprocess_onnx_image_input(
+        self, onnx_input: dict[str, np.ndarray], **kwargs
+    ) -> dict[str, np.ndarray]:
+        """
+        Preprocess the onnx input.
+        """
+        return onnx_input
```
Comment (I8dNLo) [inline fastembed/late_interaction_multimodal/onnx_multimodal_model.py:41]: Should be casted to float?

File: fastembed/late_interaction_multimodal/onnx_multimodal_model.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+        batch_size: int = 256,
+        parallel: Optional[int] = None,
+        providers: Optional[Sequence[OnnxProvider]] = None,
+        cuda: bool = False,
+        device_ids: Optional[list[int]] = None,
+        **kwargs,
+    ) -> Iterable[T]:
+        is_small = False
+
+        if isinstance(images, (str, Path, Image.Image)):
```
Comment (I8dNLo) [inline fastembed/late_interaction_multimodal/onnx_multimodal_model.py:190]: ImageInput

**Comment by I8dNLo** [comment]
To check out values for tests I use code examples from [here](https://huggingface.co/akshayballal/colpali-v1.2-merged-onnx)

File: fastembed/late_interaction_multimodal/colpali.py [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+
+class ColPali(LateInteractionMultimodalEmbeddingBase, OnnxMultimodalModel[np.ndarray]):
+    DOCUMENT_MARKER_TOKEN_ID = 2
+    QUERY_PREFIX = "Query: "
+    BOS_TOKEN = "<s>"
+    PAD_TOKEN = "<pad>"
+    QUERY_MARKER_TOKEN_ID = [2, 9413]
+    IMAGE_PLACEHOLDER_SIZE = (3, 448, 448)
+    EMPTY_TEXT_PLACEHOLDER = np.array([257152] * 1024 + [2, 50721, 573, 2416, 235265, 108])
```
Comment (joein) [inline fastembed/late_interaction_multimodal/colpali.py:44]: This is actually token ids of the following string `'<image>' * 1024 +  '<bos>Describe the image.\n'`
Could we make it nicer? It's not really readable at the moment

EVEN_ATTENTION_MASK is also not really readable, maybe instead of having this `even_attention_mask` we could assign `1030` to a constant which seems to be a bit more reasonable

**Review by joein — COMMENTED** [review]
@I8dNLo

File: tests/test_late_interaction_multimodal.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+import numpy as np
+import pytest
+
+from fastembed.late_interaction_multimodal import LateInteractionMultimodalEmbedding
+from tests.utils import delete_model_cache
+from tests.config import TEST_MISC_DIR
+from PIL import Image
+
+# vectors are abridged and rounded for brevity
+CANONICAL_COLUMN_VALUES = {
```
[the remark this answers is not in the document]
Comment (I8dNLo) [reply tests/test_late_interaction_multimodal.py:12]: I was just following our style, but you are right. It should

File: tests/test_late_interaction_multimodal.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+        token_dim = 128
+        docs = ["hello world", "flag embedding"] * 100
+        embeddings = list(model.embed_text(docs, batch_size=10, parallel=2))
+        embeddings = np.stack(embeddings, axis=0)
+
+        embeddings_2 = list(model.embed_text(docs, batch_size=10, parallel=None))
+        embeddings_2 = np.stack(embeddings_2, axis=0)
+
+        embeddings_3 = list(model.embed_text(docs, batch_size=10, parallel=0))
+        embeddings_3 = np.stack(embeddings_3, axis=0)
```
[the remark this answers is not in the document]
Comment (I8dNLo) [reply tests/test_late_interaction_multimodal.py:117]: I'll remove

File: fastembed/late_interaction_multimodal/colpali.py (hunk 5 above)
Comment (I8dNLo) [reply fastembed/late_interaction_multimodal/colpali.py:44]: That's this tokens, right. But there are some slight difference between real text input and this placeholder.
As you can see [here](https://github.com/illuin-tech/colpali/blob/f027a582c0de2aa68e1698355223d5d12e0245ec/colpali_engine/models/paligemma/colpali/processing_colpali.py#L15) original preprocessor does not add '\n' token there, while it should be added everywhere [else](https://github.com/illuin-tech/colpali/blob/f027a582c0de2aa68e1698355223d5d12e0245ec/colpali_engine/models/paligemma/colpali/processing_colpali.py#L66). So we should refactor logic of tokenize, trigger tokenizer each time (with constant output)

File: tests/test_late_interaction_multimodal.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    for model_name, expected_result in CANONICAL_IMAGE_VALUES.items():
+        print("evaluating", model_name)
+        model = LateInteractionMultimodalEmbedding(model_name=model_name)
+        result = list(model.embed_image(docs_to_embed, batch_size=2))
+
+        for value in result:
+            batch_size, token_num, abridged_dim = expected_result.shape
+            assert np.allclose(value[:token_num, :abridged_dim], expected_result, atol=1e-3)
+            break
```
Comment (joein) [inline tests/test_late_interaction_multimodal.py:61]: why do we need break here?

**Review by joein — COMMENTED** [review]
nit comments + could you please upload canonical vectors / conversion code to the dedicated repo

File: fastembed/late_interaction_multimodal/colpali.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+        return output.model_output.astype(np.float32)
+
+    def tokenize(self, documents: list[str], **_) -> list[Encoding]:
+        texts_query: list[str] = []
+        for query in documents:
+            query = self.BOS_TOKEN + self.QUERY_PREFIX + query + self.PAD_TOKEN * 10
+            query += "\n"
+
+            texts_query.append(query)
+        encoded = self.tokenizer.encode_batch(texts_query, max_length=self.QUERY_MAX_LENGTH)
```
Comment (joein) [inline fastembed/late_interaction_multimodal/colpali.py:170]: I got an exception encode_batch does not have max_length parameter
